waku/v2/api/filter.go

```diff
@@ -0,0 +1,223 @@
+	for _, t := range maps.Keys(topicMap) {
+		topicCounts[t] = 0
+	}
+	wg.Wait()
+	close(checkResults)
+	for s := range checkResults {
+		if !s.alive {
+			// Close inactive subs
+			s.sub.Close()
+			delete(apiSub.subs, s.sub.ID)
+		} else {
+			topicCounts[s.sub.ContentFilter.PubsubTopic]++
+		}
+	}
+
+	return topicCounts
+}
+
+// Attempts to resubscribe on topics that lack subscriptions
+func (apiSub *Sub) resubscribe(topicCounts map[string]int) {
+
+	// Delete healthy topics
+	for t, cnt := range topicCounts {
+		if cnt == apiSub.Config.MaxPeers {
+			delete(topicCounts, t)
+		}
+	}
+
+	if len(topicCounts) == 0 {
+		// All topics healthy, return
+		return
+	}
+
+	// Re-subscribe asynchronously
+	newSubs := make(chan []*subscription.SubscriptionDetails)
+
+	for t, cnt := range topicCounts {
+		cFilter := protocol.ContentFilter{PubsubTopic: t, ContentTopics: apiSub.ContentFilter.ContentTopics}
+		go func(count int) {
+			subs, _ := apiSub.subscribe(cFilter, apiSub.Config.MaxPeers-count)
+			newSubs <- subs
+		}(cnt)
+	}
+
+	cnt := 0
+	apiSub.log.Debug("resubscribe(): before range newSubs")
+	for subs := range newSubs {
+		cnt++
+		if subs != nil {
+			apiSub.multiplex(subs)
+		}
+		if cnt == len(topicCounts) {
+			// Received all subscription results
+			break
+		}
+	}
+	apiSub.log.Info("checkAliveness(): close(newSubs)")
+	close(newSubs)
+}
+
+func (apiSub *Sub) subscribe(contentFilter protocol.ContentFilter, peerCount int) ([]*subscription.SubscriptionDetails, error) {
+	// Low-level subscribe, returns a set of SubscriptionDetails
+	options := make([]filter.FilterSubscribeOption, 0)
+	options = append(options, filter.WithMaxPeersPerContentFilter(int(peerCount)))
+	for _, p := range apiSub.Config.Peers {
+		options = append(options, filter.WithPeer(p))
+	}
+	subs, err := apiSub.wf.Subscribe(apiSub.ctx, contentFilter, options...)
+
+	if err != nil {
+		if len(subs) > 0 {
+			// Partial Failure, for now proceed as we don't expect this to happen wrt specific topics.
+			// Rather it can happen in case subscription with one of the peer fails.
+			// This can further get automatically handled at resubscribe,
+			apiSub.log.Error("partial failure in Filter subscribe", zap.Error(err))
+			return subs, nil
+		}
+		// In case of complete subscription failure, application or user needs to handle and probably retry based on error
+		// TODO: Once filter error handling indicates specific error, this can be addressed based on the error at this layer.
+		return nil, err
+	}
+
+	return subs, nil
+}
+
+func (apiSub *Sub) multiplex(subs []*subscription.SubscriptionDetails) {
+
+	// Multiplex onto single channel
+	// Goroutines will exit once sub channels are closed
+	for _, subDetails := range subs {
+		apiSub.subs[subDetails.ID] = subDetails
+		go func(subDetails *subscription.SubscriptionDetails) {
+			apiSub.log.Info("New multiplex", zap.String("subID", subDetails.ID))
+			for env := range subDetails.C {
+				apiSub.DataCh <- env
+			}
```

Comment:
Would it be possible for the code to run into an error attempting to write to `DataCh` after it's closed? or this will never happen because all the `subDetails.C` channels will be closed before `close(apiSub.DataCh)` is executed?

Reply:
Yup, as you pointed out all subs are closed which internally close the channels beofre the dataCh is closed. 
So, it should be fine and will not run into this issue.